src/goto-instrument/contracts/instrument_spec_assigns.cpp

```diff
@@ -513,3 +682,117 @@ void instrument_spec_assignst::invalidate_heap_and_spec_aliases(
   for(const auto &pair : from_heap_alloc)
     invalidate_car(pair.second, freed_car, dest);
 }
+
+/// Returns true iff an `ASSIGN lhs := rhs` instruction must be instrumented.
+bool instrument_spec_assignst::must_check_assign(
+  const goto_programt::const_targett &target,
+  skip_function_paramst skip_function_params,
+  const optionalt<cfg_infot> cfg_info_opt)
+{
+  if(
+    const auto &symbol_expr =
+      expr_try_dynamic_cast<symbol_exprt>(target->assign_lhs()))
+  {
+    if(
+      skip_function_params == skip_function_paramst::No &&
+      ns.lookup(symbol_expr->get_identifier()).is_parameter)
+      return true;
+
+    if(cfg_info_opt.has_value())
+      return !cfg_info_opt.value().is_local(symbol_expr->get_identifier());
+  }
+
+  return true;
+}
+
+/// Returns true iff a `DECL x` must be added to the local write set.
+///
+/// A variable is called 'dirty' if its address gets taken at some point in
+/// the program.
+///
+/// Assuming the goto program is obtained from a structured C program that
+/// passed C compiler checks, non-dirty variables can only be assigned to
+/// directly by name, cannot escape their lexical scope, and are always safe
+/// to assign. Hence, we only track dirty variables in the write set.
+bool instrument_spec_assignst::must_track_decl(
+  const goto_programt::const_targett &target,
+  const optionalt<cfg_infot> &cfg_info_opt) const
+{
+  if(cfg_info_opt.has_value())
+    return cfg_info_opt.value().is_not_local_or_dirty_local(
+      target->decl_symbol().get_identifier());
+
+  // Unless proved non-dirty by the CFG analysis we assume it is dirty.
+  return true;
+}
+
+/// Returns true iff a `DEAD x` must be processed to upate the local write set.
+/// The conditions are the same than for tracking a `DECL x` instruction.
+bool instrument_spec_assignst::must_track_dead(
+  const goto_programt::const_targett &target,
+  const optionalt<cfg_infot> &cfg_info_opt) const
+{
+  // Unless proved non-dirty by the CFG analysis we assume it is dirty.
+  if(!cfg_info_opt.has_value())
+    return true;
+
+  return cfg_info_opt.value().is_not_local_or_dirty_local(
+    target->dead_symbol().get_identifier());
+}
+
+void instrument_spec_assignst::instrument_assign_statement(
+  goto_programt::targett &instruction_it,
+  goto_programt &program,
+  optionalt<cfg_infot> &cfg_info_opt) const
+{
+  auto lhs = instruction_it->assign_lhs();
+  lhs.add_source_location() = instruction_it->source_location();
+  goto_programt payload;
+  check_inclusion_assignment(lhs, cfg_info_opt, payload);
+  insert_before_swap_and_advance(program, instruction_it, payload);
+}
+
+void instrument_spec_assignst::instrument_call_statement(
+  goto_programt::targett &instruction_it,
+  goto_programt &body,
+  optionalt<cfg_infot> &cfg_info_opt)
+{
+  INVARIANT(
+    instruction_it->is_function_call(),
+    "The first argument of instrument_call_statement should always be "
+    "a function call");
```

Comment:
`PRECONDITION` may be more specific than the most generic `INVARIANT` here.

Reply:
fixed